corehq/apps/user_importer/importer.py

```diff
@@ -582,35 +579,35 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
                     # Passing use_primary_db=True because of https://dimagi-dev.atlassian.net/browse/ICDS-465
                     user.get_django_user(use_primary_db=True).check_password(password)
 
-                for group in domain_info.group_memoizer.by_user_id(user.user_id):
-                    if group.name not in group_names:
-                        group.remove_user(user)
+                error = commcare_user_importer.update_user_groups(domain_info, group_names)
+                if error:
+                    error_message = f"{user.user_id}: {error}"
+                    ret['errors'].append(_(error_message))
+                try:
+                    for domain_info in domain_info_by_domain.values():
```

Comment:
I think this should probably stay in the `finally` block to prevent re-saving the groups for every user row processed. I guess the tricky thing now is that the log is saved in the loop but the groups are saved out of the loop.

One option would be to make the memoizer smarter so that it knows which groups have been updated and need to be saved. That way we could do the save in the loop and avoid having to save all the groups every time. It would still be more expensive than doing the save at the end but we'd have better consistency between the data and the logs.

Reply:
I added a set to track group changes to the memoizer. It's populated for each user in the loop by `update_user_groups`,  saved and then cleared for the next user.